json-rpc/src/stream_rpc/subscriptions.rs

```diff
@@ -0,0 +1,258 @@
+
+#[derive(Clone, Debug)]
+pub struct SubscriptionConfig {
+    pub fetch_size: u64,
+    pub poll_interval_ms: u64,
+    pub queue_size: usize,
+}
+
+type JitterBackoff = Map<ExponentialBackoff, fn(Duration) -> Duration>;
+
+pub fn create_backoff(poll_interval_ms: u64) -> JitterBackoff {
+    ExponentialBackoff::from_millis(poll_interval_ms)
+        .factor(1000)
+        .max_delay(tokio::time::Duration::from_millis(poll_interval_ms * 10))
+        .map(jitter)
+}
+
+#[derive(Clone)]
+pub struct SubscriptionHelper {
+    pub db: Arc<dyn DbReader>,
+    pub client: ClientConnection,
+    pub jsonrpc_id: Id,
+    pub method: Method,
+    backoff: JitterBackoff,
+}
+
+impl SubscriptionHelper {
+    pub fn new(
+        db: Arc<dyn DbReader>,
+        client: ClientConnection,
+        jsonrpc_id: Id,
+        method: Method,
+    ) -> Self {
+        let poll_interval_ms = client.config.poll_interval_ms;
+        Self {
+            db,
+            client,
+            jsonrpc_id,
+            method,
+            backoff: create_backoff(poll_interval_ms),
+        }
+    }
+    /// This is to be one of the first things that a new subscription task must do, to let the client
+    /// know that the subscription has been created (some streams may be very sparse, or start with 0 items)
+    pub async fn send_ok(&self) {
+        counters::SUBSCRIPTION_OKS
+            .with_label_values(&[
+                self.client.connection_context.transport.as_str(),
+                &self.method.as_str(),
+                counters::RpcResult::Success.as_str(),
+                self.client.connection_context.sdk_info.language.as_str(),
+                &self.client.connection_context.sdk_info.version.to_string(),
+            ])
+            .inc();
+        self.client
+            .send_success(
+                self.jsonrpc_id.clone(),
+                &SubscribeResult::ok(self.db.get_latest_version().unwrap_or(0)),
+            )
+            .await
+            .ok();
+    }
+
+    pub fn reset_backoff(&mut self) {
+        self.backoff = create_backoff(self.client.config.poll_interval_ms)
+    }
+
+    pub async fn sleep_wiggled(&mut self) {
+        let sleepy_time = self.backoff.next().unwrap();
+        debug!(
+            "Client#{} Sleeping for {:?}ms as no results for {}",
+            self.client.id,
+            sleepy_time,
+            self.method.as_str()
+        );
+        tokio::time::sleep(sleepy_time).await;
+    }
+}
+
+pub trait Subscription<ParamType>: Send + Sync + Clone + 'static
+where
+    ParamType: Send + Sync + Serialize + Clone,
+{
+    fn init(&mut self, helper: &SubscriptionHelper) -> Result<(), JsonRpcError>;
+    fn next(&self, helper: &SubscriptionHelper) -> Vec<ParamType>;
+    fn on_send(&mut self, item: &ParamType);
+
+    fn run(mut self, mut helper: SubscriptionHelper) -> Result<JoinHandle<()>, JsonRpcError> {
+        self.init(&helper)?;
+
+        Ok(tokio::spawn(async move {
```

Comment:
I'm curious how this `tokio::spawn` closure relates to the `helper#ClientConnection` spawn routine used to manage raw messages.

Suggestion:
At a high level, it would be insightful to know how this PubSub implementation fans out and in. I believe explicitly capturing it in comments or dialog can drive out a leaner or more readable implementation.

example: ClientConnection is on a single tokio routine that's created by ConnectionManager responsible for fan out based on new connection. The SubscriptionHelper has it's own goroutine because ....

Reply:
Does this help explain it? I added it to the readme:
```
### Task Fanout
1. Each `ConnectionManager` spawns two `tokio::Task` for each connection (one each for monitoring incoming/outgoing channels).
2. Each `ClientConnection` then launches one `Task` per `Subscription`, which run in parallel.
```

Reply:
WRT to 2, can we have many subscriptions to one ClientConnection? Client makes multiple subscription calls with different filters? If so, have we considered one task to handle that? It would fan in on the `ConnectionManager#outgoing channel` anyways. I am leaning on your insight into the perf savings of a single task. Perhaps the other end of the subscription results in a connection onchain that we can save.

Reply:
A client can make as many subscriptions as they want- i.e one to `Events` and one to `Transactions`, 10 to `Events`, etc.

I believe using a single task would introduce a very significant amount of complexity, and if possible I'd like to save the benchmarking and potential performance improvements for a V2: this API is not meant to be exposed to the general internet, and so it's much easier for anyone trying to use this at a very large scale to spin up another full node   than performance improvements (to which there is a ceiling anyway). Does that make sense?